### PR TITLE
add qr_backward functionality for wide case

### PR DESCRIPTION
**Pull request by rlucas7**
Unblocks implementation of #27036. Note that this PR ***does not*** fix #{27036}. 
Currently QR decomposition only has support for square and tall (a.k.a. skinny) case. 
This PR adds functionality for wide A matrix/tensors, includes 3 unit tests for the new case
and restructures the `qr_backward` method to use the same Walther method as a helper. 

cc @albanD @t-vi 

I don't have a gpu machine so haven't tested on cuda but everything passes on my local machine in cpu. 

The basic idea of the PR is noted in the comments in the `Functions.cpp` file but I'll note here too for clarity: 

let  be a matrix and  then partition  as  
and take QR of  and call that one 
 the  here from  is the same as the  from  on entire  matrix. Then transform  with the  rotation got from  to get  now  and similarly for the grads of each piece, e.g. if  is  `grad_A` then 
 and  and then 
 and 
 is the `narrow()` of `grad_R`. 
 is calculated very similar to the original Walther formula (exactly the same in the tall and square cases) but is slightly modified here for wide case matrices.

### CODE REVIEW

**Comment by t-vi** [comment]
@lucas7 Thank you for working on this! Looks good overall, thank you for throwing in tests, too.

I'll have to look at the maths some more, thank you for the explanation. In the meantime, could you fix the tabs, please?

**Comment by albanD** [comment]
The diff is not great indeed, maybe the tabs. You can replace them and push again to make the linter happy.

Also is it correct that `square_deep_case_backward` corresponds exactly to the previous function? Or did you change things in there?

File: tools/autograd/templates/Functions.cpp
Comment (rlucas7) [inline tools/autograd/templates/Functions.cpp:2051]: so what I mean by exactly the same is that when you call `square_deep_case_backward` in this block is the same result as previous implementation.

**Comment by rlucas7** [comment]
> Thanks for the PR!
> 
> The diff is not great indeed, maybe the tabs. You can replace them and push again to make the linter happy.

Yeah that's what I get for cutting a PR so late at night. Will remove tabs on update. 
 
> Also is it correct that `square_deep_case_backward` corresponds exactly to the previous function? Or did you change things in there?

Yeah sorry I wasn't communicating clearly, what I meant with the comment is that in the case that the input matrix is tall or square, the calculation is the same. I left a comment inline in the PR where I was referring to, apologies for the confusion.

**Comment by rlucas7** [comment]
> Thanks for the PR!
> 
> The diff is not great indeed, maybe the tabs. You can replace them and push again to make the linter happy.

Yeah that's what I get for cutting a PR so late at night. Will remove tabs on update. 
 
> Also is it correct that `square_deep_case_backward` corresponds exactly to the previous function? Or did you change things in there?

Yeah sorry I wasn't communicating clearly, what I meant with the comment is that in the case that the input matrix is tall or square, the calculation is the same. There was a small changes towards the end of the lambda to account for functionality of wide. 

> @lucas7 Thank you for working on this! Looks good overall, thank you for throwing in tests, too.
> 
> I'll have to look at the maths some more, thank you for the explanation. In the meantime, could you fix the tabs, please?

Oops, wrong alias, I'm ***rlucas7*** but whoever lucas7 is, he's got an awesome avatar :)

**Comment by rlucas7** [comment]
I ran `flake8 tools/autograd/templates/Functions.cpp` resolved all the tabs and other `EXXX` issues that came up in lines I touched, ≈LIC 1973-LIC 2090. Also did a `mv .flake8 .otherfilename` and re-ran flake8, the remaining issues on the lines in this file are all E501 but the `.flake8` file sets the line line to 120 instead of 80 so should be fine. There are still some `flake8` errors when I run `flake8 tools/autograd/templates/Functions.cpp` locally, but these are LIC 1 - LIC 31 which I didn't touch. If you'd like to me to resolve those style/flake8 issues within this PR I can but they aren't related to the changes so I chose to omit unless instructed o/w, hope that is ok.

Also, not sure if you prefer me to do a rebase with changes, I usually wait until all comments addressed to rebase and squash and this one was still addressing the changes to get the builds to green.

**Comment by t-vi** [comment]
@rlucas7 sorry about the username. You want git-clang-format or so for C++, not flake8.

**Comment by t-vi** [comment]
Relative to the organisation of the code: I wonder why we could drop the lambda here, it seems to me that we usually just use `_qr_backward_square_tall` or so for our helper functions in Functions.cpp (e.g. `_euclidean_dist_backward`). The functions aren't exported, so I don't see the immediate benefit of having the lambda. I think this would make it a bit more clearer what is old and new, too, as the old function could just continue to live on as the helper.
What do you think?

**Comment by albanD** [comment]
All lint is fine now :)
I would agree that if we can make the diff cleaner it would be great. If you cannot do it it's fine as well. Will just be a bit longer to review.

File: tools/autograd/templates/Functions.cpp
Comment (rlucas7) [inline tools/autograd/templates/Functions.cpp:2052]: From here down is the control flow for the wide case, which is the new stuff.

File: tools/autograd/templates/Functions.cpp
Comment (rlucas7) [inline tools/autograd/templates/Functions.cpp]: This part might be a little confusing b/c of the diff these lines are all kept in the lambda. Github is splittting these changes across 3 ***patches*** so they git rendered separately the the browser, if you look above at line 2012 that was removed the `grad_R.defined()` control flow is preserved, also the `grad_Q.defined()` control flow is preserved but b/c of the multiple patches the changed lines are getting "striped" so it doesn't appear that they are the same between old and this PR but these lines ***are*** preserved.

File: tools/autograd/templates/Functions.cpp
Comment (rlucas7) [inline tools/autograd/templates/Functions.cpp:2059]: This comment is using the python `@` operator for the (Euclidean) inner product. If this is confusing as written I can replace with the equivalent operator for the `Tensor` class.

File: tools/autograd/templates/Functions.cpp
Comment (rlucas7) [inline tools/autograd/templates/Functions.cpp:2034]: This line is now inside the lambda, it is the last line of the lambda, the (added) LIC 2037

**Comment by rlucas7** [comment]
> Relative to the organisation of the code: I wonder why we could drop the lambda here, it seems to me that we usually just use `_qr_backward_square_tall` or so for our helper functions in Functions.cpp (e.g. `_euclidean_dist_backward`). The functions aren't exported, so I don't see the immediate benefit of having the lambda. 

I was also a bit confused on the choice of a `_*` method versus a lambda. Unfortunately, the `Functions.cpp` file isn't internally consistent on this (or at least it has some logic I didn't grasp). There are cases of helper methods like you mention and also examples of lambdas, which are used throughout the matrix autodiff methods. For example,`det_backward` uses 2 lambdas:
One for the singular case
https://github.com/pytorch/pytorch/blob/26d58503c220a709117582f48cff77a6fc418d6d/tools/autograd/templates/Functions.cpp#L2040
And one for the non-singular case
https://github.com/pytorch/pytorch/blob/26d58503c220a709117582f48cff77a6fc418d6d/tools/autograd/templates/Functions.cpp#L2047

and a similar structure is followed for the other matrix backprop methods `slogdet_backward`
https://github.com/pytorch/pytorch/blob/26d58503c220a709117582f48cff77a6fc418d6d/tools/autograd/templates/Functions.cpp#L2136
and for `logdet_backward` 
https://github.com/pytorch/pytorch/blob/26d58503c220a709117582f48cff77a6fc418d6d/tools/autograd/templates/Functions.cpp#L2088
plus those are closer in the `Functions.cpp` file to where `qr_backward` sits, which is why I chose the lambda approach ultimately. 
If function names aren't exported anywhere then the only benefit would be if the function is used in another function in the `Functions.cpp`. I didn't see any case of that in the `Functions.cpp`.

>I think this would make it a bit more clearer what is old and new, too, as the old function could just continue to live on as the helper.
> What do you think?

I don't think it will make it clearer what is old and what is new. I initially had written the function as a separate function to keep it simpler. IIRC the only changes I made when moving it into the lambda was changing some of the matrix overwrites to reduce memory use and to handle  cases for `grad_Q.defined()` and/or `grad_R.defined()`.  Now those changes weren't in the lambda itself but in the `qr_backward` control flow for the new wide case, I've left an inline note where that part of the changeset occurs. 

I think the confusion is around the way github is striping the lambda refactor in the diff as 3 patches, obfuscating that this is really a lift and shift into a lambda. For this I also left some inline comments to help clarify what's going on in terms of the changes. 

Maybe take a look at that and let me know if it's still unclear and you want to pull it out as a separate function? 

If you both still prefer it as a separate `_*` helper after looking at comments I can change it to be written that way instead. 

Overall I think it might help for me to enumerate changes that are made to `qr_backward` in this PR:
1. refactor original `qr_backward` calculation into a lambda
2. add support for `qr_backward` when `self` (or A as I call it in lambda) is wide (more columns than rows)
Note this change includes some overwrites of Tensors to use less memory in the calculations. 
3. additional handling for when the `grad_R` and/or `grad_Q` are (or are not) defined. 

Also would it help if I shared some notes on the Maths here for the wide case?  

Using Github's latex injection is pretty clunky but I could write up the algebra for the wide case in latex and share it via email. Some of the other SciPy devs and I have done that before on some random variate generators and found it helpful to understand why things were coded as they were.

**Comment by rlucas7** [comment]
Hello, friendly inquiry, anything further needed from my end at this time-any updates?

File: tools/autograd/templates/Functions.cpp
Comment (albanD) [inline tools/autograd/templates/Functions.cpp:2086]: Actually this does not.
`foo = bar` does not write into `foo` inplace. It just frees the old content of foo and put a copy (talking about shared_ptr like object here, so no content copy) of bar into it.
Can you just update the comment?

In general, I would advise against doing inplace changes here as it would most likely prevent double backward from working.
In this case, grad_Y is actually a view of the provided grads that should *never* be modified inplace.

File: torch/testing/_internal/common_methods_invocations.py
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:792]: I think you actually need to make the 5th element, which is an empty tuple right now into `(True,)` for the gradients to be checked.
Can you try to make that change and make sure the gradcheck passes.

Also you can activate them for the square case that was not tested before :o

**Review by albanD — COMMENTED** [review]
Hey,
Sorry for the delay.
The code change looks good to me. I added just small comments for formatting and comments.

But looking at the test, I think the old specification was not correct and not checking the gradients. It would be nice if you could re-activate that for all the settings and make sure the formula works fine!

File: tools/autograd/templates/Functions.cpp
Comment (rlucas7) [reply tools/autograd/templates/Functions.cpp:2086]: > Actually this does not.
> `foo = bar` does not write into `foo` inplace. It just frees the old content of foo and put a copy (talking about shared_ptr like object here, so no content copy) of bar into it.
> Can you just update the comment?

I will remove the line from the comments. 

I think we also want to remove the overwrite, correct?

Thanks for catching that, is this what you're referring to? 
https://github.com/pytorch/pytorch/blob/888ae1b3d8bfc294ccace3c3c7b4b55046fb7d9e/aten/src/ATen/native/LinearAlgebra.cpp#L679-L680

IIUC, I will add a `grad_V` to the control flow and removed the overwriting so it will read:
`grad_Y = at::matmul(q, grad_V);` on LIC 2087
 (changing the `grad_Y` entries above to read `grad_V`.
> In general, I would advise against doing inplace changes here as it would most likely prevent double backward from working.
> In this case, grad_Y is actually a view of the provided grads that should _never_ be modified inplace.

File: tools/autograd/templates/Functions.cpp
Comment (rlucas7) [reply tools/autograd/templates/Functions.cpp:2040]: yes, good idea, that will make it clearer

**Comment by rlucas7** [comment]
> Hey,
> Sorry for the delay.

No worries @albanD thanks for your time and thorough review. 

> The code change looks good to me. I added just small comments for formatting and comments.
> 
> But looking at the test, I think the old specification was not correct and not checking the gradients. It would be nice if you could re-activate that for all the settings and make sure the formula works fine!

I wasn't clear on this one after doing some repo spelunking, I left a couples links at the inline comment of what I found, doesn't seem that the 5th tuple is used? 

The units tests do seem to pass both with and without them though.

File: tools/autograd/templates/Functions.cpp
Comment (rlucas7) [reply tools/autograd/templates/Functions.cpp:2059]: Hmm, an assumption of the current (tall and square) implementation is that from the forward pass, `k=min(m,n)`, where `A` is an `m x n` matrix, that has `rank(A) = k`, `Q` is `m x k` and `R` is `k x n`.
 
This PR ***does not*** change those assumptions. 

If it is clearer, I can rephrase the comment to include the word/phrase *assumption*? 

FWIW you seem to be referring to the case `l<k=min(m,n)`, where `l=rank(A)`.

Comment (nikitaved) [reply tools/autograd/templates/Functions.cpp:2059]: No, unfortunately Q is not going to be `m x k`, it is going to be either `m x min(m, n)` or `m x m`. The issue is with R, it is going to be full-rank if and only if first columns of `A` are linearly-independent.

In the comment you state that for `A = (X | Y)` `X` is of full rank, which is not the case in general. Note that current QR is without pivoting, so it  performs a factorization `A = QR` where `R` is upper-triangular, but it does not mean that `rank(R[:, 1:l]) = l for l <= rank(A)`.

Comment (nikitaved) [reply tools/autograd/templates/Functions.cpp:2059]: As a counterexample just take `A` of rank 1.

Comment (nikitaved) [reply tools/autograd/templates/Functions.cpp:2059]: Yes, put a comment that it is assumed that `rank(A) = n`. But, still, how do you know that it is exactly first `n` columns of `A` are linearly-independent?

Comment (nikitaved) [reply tools/autograd/templates/Functions.cpp:2059]: > Hmm, an assumption of the current (tall and square) implementation is that from the forward pass, `k=min(m,n)`, where `A` is an `m x n` matrix, that has `rank(A) = k`, `Q` is `m x k` and `R` is `k x n`.

The documentation says nothing about the rank of an input as far as I can see.

**Comment by albanD** [comment]
@nikitaved what about we add a warning to the doc specifying the current limitation of the backward for qr and add this improvement (that has the same limitations).
And in a future PR, after when we update the QR forward, we can update the formula here to perform the right thing?

**Comment by rlucas7** [comment]
>@albanD wrote: 
>> @nikitaved what about we add a warning to the doc specifying the current limitation of the backward for qr and add this improvement (that has the same limitations).
>> And in a future PR, after when we update the QR forward, we can update the formula here to perform the right thing?

>@nikitaved wrote: 
>>@albanD , sure, makes sense. I can actually update the documentation.

It seems like the tril solver issue and the dgeqp3 request are separate issues.
I agree the suggestion of @albanD seems like a reasonable path forward.

AFAICS the only change here with a RRQR (on forward pass) would be a right multiply by a permutation matrix `P` before doing gradient calculations. 
Then after the gradients are calculations post multiply the returned `grad_A` via `P^T`, the inverse of `P`, to move the columns back to their original arrangements. 
This would be a very small change to the existing codes (once dgeqp3 is brought into pytorch in forward pass).

The use of RRQR would not change the fact that the equation requires `rank(A)=k=min(m,n)`. 

@albanD it looks like there are some conflicts on this branch, causing test failures in CI. 
Difficult for me to determine, is a `git rebase -i master` required here to get the CI to green, or is there something more required? 

I make this guess from ctrl-f in this file for error:

https://dr.pytorch.org/api/view-log-full?build_id=129136002

seems to indicate merge conflicts, the other windows builds seem to have network connections issues (unrelated I assume).

**Comment by albanD** [comment]
Hi,

@nikitaved already sent a PR for the doc so that's good.

Yes, can you rebase on top of master to make CI happy please?

**Review by albanD — APPROVED** [review]
LGTM 
Will wait on the updated CI signal before merging.

**Comment by dvirginz** [comment]
@rlucas7 Thank you very much for the contribution!
Just to make sure, should this make 'lstsq' differentiable?
I'm using the nightly version to torch ('1.7.0.dev20200902+cu101`), and still having the `the derivative for 'lstsq' is not implemented` error.
Thanks!

**Comment by rlucas7** [comment]
This PR is ***part*** of what you would need to make lstsq autograd work. 

There are a couple other things that would also need to be in place before 
you could do lstsq backward pass. 

-Lucas Roberts

**Pull request closed without merging**
